sklearn/metric_learning/nca.py

```diff
@@ -0,0 +1,280 @@
+from ..base import BaseEstimator, TransformerMixin
+from ..utils.validation import check_is_fitted
+
+import numpy as np
+import scipy as sp
+import scipy.misc
+import scipy.optimize
+import time
+
+import sys
+
+
+def nca_vectorized_oracle(X, y, n_components, loss, threshold=0.0):
+    dx = X[:, np.newaxis, :] - X[np.newaxis, :, :]  # n_samples x n_samples x n_features
+    outer = dx[:, :, np.newaxis, :] * dx[:, :, :, np.newaxis]  # n_samples x n_samples x n_features x n_features
```

Comment:
Could you check whether this full expansion is necessary? Looking at the formulas for the gradient, it seems possible to bring `L` inside all members of the sum. Assuming that `L.shape[0] <= L.shape[1]` (is there a usecase for the opposite?), this has the potential to decrease memory and cpu cycles on this outer expansion.